File: SuperiorPackGroup/Reports/ProductionSummaryXtraReport.vb
Imports System.Linq
Imports DevExpress
Imports DXDAL.SPGData

Public Class ProductionSummaryXtraReport

    Private m_BreakMinutes As Long
    Private m_Minutes As Integer
    Private m_TimeDiffStartToStop As Long
    Private m_MinPerHour As Integer = 50
    Private m_CurrentProduction As Production

    Private Sub ProductionSummaryXtraReport_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles MyBase.BeforePrint

        'm_MinPerHour = My.Forms.MainXtraForm.MinPerHOur

    End Sub

    Private Sub totalMinutes_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles totalMinutes.BeforePrint

        Dim hours As Integer = CInt(totalHours.Summary.GetResult.ToString.Substring(0, totalHours.Summary.GetResult.ToString.IndexOf(":"))) * 60
        Dim minutes As Integer = CInt(totalHours.Summary.GetResult.ToString.Substring(totalHours.Summary.GetResult.ToString.IndexOf(":") + 1)) + hours
        totalMinutes.Text = CStr(((minutes - m_BreakMinutes) / 60) * m_MinPerHour)

    End Sub

    Private Sub averageActualSpeed_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles averageActualSpeed.BeforePrint

        Dim hours As Integer = CInt(totalHours.Summary.GetResult.ToString.Substring(0, totalHours.Summary.GetResult.ToString.IndexOf(":"))) * 60
        Dim minutes As Integer = CInt(totalHours.Summary.GetResult.ToString.Substring(totalHours.Summary.GetResult.ToString.IndexOf(":") + 1)) + hours
        averageActualSpeed.Text = Format(CDbl(AverageActualSpeedH.Summary.GetResult) / CType((((minutes - m_BreakMinutes) / 60) * m_MinPerHour), Integer), "0.00")

    End Sub

    Private Sub totalPacksPerMin_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles totalPacksPerMin.BeforePrint

        Dim hours As Integer = CInt(totalHours.Summary.GetResult.ToString.Substring(0, totalHours.Summary.GetResult.ToString.IndexOf(":"))) * 60
        Dim minutes As Double = ((CInt(totalHours.Summary.GetResult.ToString.Substring(totalHours.Summary.GetResult.ToString.IndexOf(":") + 1)) + hours) / 60) * m_MinPerHour
        totalPacksPerMin.Text = Format(CDbl(totalQtyProduced.Summary.GetResult) / (minutes - m_BreakMinutes), "0.00")

    End Sub

    Private Sub totalMinutesLabel_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles totalMinutesLabel.BeforePrint

        totalMinutesLabel.Text = String.Format("Total Minutes ({0}/Hour)", m_MinPerHour)

    End Sub

    Private Sub totalHours_SummaryGetResult(ByVal sender As System.Object, ByVal e As DevExpress.XtraReports.UI.SummaryGetResultEventArgs) Handles totalHours.SummaryGetResult

        e.Result = String.Format("{0}:{1}", Strings.Format(m_Minutes \ 60, "##############0"), Strings.Format((m_Minutes Mod 60) \ 1, "00"))
        e.Handled = True

    End Sub

    Private Sub totalHours_SummaryReset(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles totalHours.SummaryReset

        m_Minutes = 0

    End Sub

    Private Sub totalHours_SummaryRowChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles totalHours.SummaryRowChanged

        m_Minutes += CInt(DateDiff(DateInterval.Minute, Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStart")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStop"))) - _
                        (BreakTimeBLL.GetBreakMinutes(Convert.ToDateTime(GetCurrentColumnValue("ProdMainDate")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStart")), _
                                                      Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStop")))))

    End Sub

    Private Sub Detail_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles Detail.BeforePrint

        m_TimeDiffStartToStop = DateDiff(DateInterval.Minute, Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStart")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStop")))
        m_BreakMinutes = BreakTimeBLL.GetBreakMinutes(Convert.ToDateTime(GetCurrentColumnValue("ProdMainDate")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStart")), _
                                                      Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStop")))

        m_CurrentProduction = Xpo.XpoDefault.Session.GetObjectByKey(Of Production)(CInt(GetCurrentColumnValue("ProdMainID")))

        Dim MachineStandard As ItemMachineStandards
        If m_CurrentProduction IsNot Nothing Then
            MachineStandard = m_CurrentProduction.ProdMainItemID.MachineStandards.Where(Function(IMS) IMS.Machine.MachineLineID = m_CurrentProduction.ProdMainMachineLine.MachineLineID).FirstOrDefault()
            If MachineStandard IsNot Nothing Then
                If MachineStandard.MinutesPerShift > 0 Then
                    m_MinPerHour = CInt(MachineStandard.MinutesPerShift / 7.16667)
                Else If m_CurrentProduction.ProdMainItemID.MinutesPerShift > 0
                    m_MinPerHour = CInt(m_CurrentProduction.ProdMainItemID.MinutesPerShift / 7.1667)
                Else
                    m_MinPerHour = MainXtraForm.MinPerHOur
                End If  
            Else If m_CurrentProduction.ProdMainItemID.MinutesPerShift > 0
                m_MinPerHour = CInt(m_CurrentProduction.ProdMainItemID.MinutesPerShift / 7.1667)              
            Else
                m_MinPerHour = MainXtraForm.MinPerHOur
            End If
        Else
            m_MinPerHour = MainXtraForm.MinPerHOur
        End If

    End Sub

End Class

File: SuperiorPackGroup/Reports/ProductionSummaryXtraReport.cs
//INSTANT C# NOTE: Formerly VB project-level imports:
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;

using System.Linq;
using DevExpress;
using DXDAL.SPGData;

namespace SuperiorPackGroup
{
	public partial class ProductionSummaryXtraReport
	{

		public ProductionSummaryXtraReport()
		{
			InitializeComponent();
		}

		private long m_BreakMinutes;
		private int m_Minutes;
		private long m_TimeDiffStartToStop;
		private int m_MinPerHour = 50;
		private Production m_CurrentProduction;

		private void ProductionSummaryXtraReport_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			//m_MinPerHour = My.Forms.MainXtraForm.MinPerHOur

		}

		private void totalMinutes_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			int hours = int.Parse(totalHours.Summary.GetResult().ToString().Substring(0, totalHours.Summary.GetResult().ToString().IndexOf(":"))) * 60;
			int minutes = int.Parse(totalHours.Summary.GetResult().ToString().Substring(totalHours.Summary.GetResult().ToString().IndexOf(":") + 1)) + hours;
			totalMinutes.Text = (((minutes - m_BreakMinutes) / 60.0) * m_MinPerHour).ToString();

		}

		private void averageActualSpeed_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			int hours = int.Parse(totalHours.Summary.GetResult().ToString().Substring(0, totalHours.Summary.GetResult().ToString().IndexOf(":"))) * 60;
			int minutes = int.Parse(totalHours.Summary.GetResult().ToString().Substring(totalHours.Summary.GetResult().ToString().IndexOf(":") + 1)) + hours;
			averageActualSpeed.Text = (Convert.ToDouble(AverageActualSpeedH.Summary.GetResult()) / Convert.ToInt32((((minutes - m_BreakMinutes) / 60.0) * m_MinPerHour))).ToString("0.00");

		}

		private void totalPacksPerMin_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			int hours = int.Parse(totalHours.Summary.GetResult().ToString().Substring(0, totalHours.Summary.GetResult().ToString().IndexOf(":"))) * 60;
			double minutes = ((int.Parse(totalHours.Summary.GetResult().ToString().Substring(totalHours.Summary.GetResult().ToString().IndexOf(":") + 1)) + hours) / 60.0) * m_MinPerHour;
			totalPacksPerMin.Text = (Convert.ToDouble(totalQtyProduced.Summary.GetResult()) / (minutes - m_BreakMinutes)).ToString("0.00");

		}

		private void totalMinutesLabel_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			totalMinutesLabel.Text = string.Format("Total Minutes ({0}/Hour)", m_MinPerHour);

		}

		private void totalHours_SummaryGetResult(System.Object sender, DevExpress.XtraReports.UI.SummaryGetResultEventArgs e)
		{

			e.Result = string.Format("{0}:{1}", m_Minutes / 60, (m_Minutes % 60) / 1);
			e.Handled = true;

		}

		private void totalHours_SummaryReset(System.Object sender, EventArgs e)
		{

			m_Minutes = 0;

		}

		private void totalHours_SummaryRowChanged(System.Object sender, EventArgs e)
		{

			m_Minutes += Convert.ToInt32(DateHelper.DateDiff(DateHelper.DateInterval.Minute, Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStart")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStop"))) - (BreakTimeBLL.GetBreakMinutes(Convert.ToDateTime(GetCurrentColumnValue("ProdMainDate")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStart")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStop")))));

		}

		private void Detail_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			m_TimeDiffStartToStop = DateHelper.DateDiff(DateHelper.DateInterval.Minute, Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStart")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStop")));
			m_BreakMinutes = BreakTimeBLL.GetBreakMinutes(Convert.ToDateTime(GetCurrentColumnValue("ProdMainDate")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStart")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStop")));

			m_CurrentProduction = DevExpress.Xpo.XpoDefault.Session.GetObjectByKey<Production>(Convert.ToInt32(GetCurrentColumnValue("ProdMainID")));

			ItemMachineStandards MachineStandard = null;
			if (m_CurrentProduction != null)
			{
				MachineStandard = m_CurrentProduction.ProdMainItemID.MachineStandards.Where((IMS) => IMS.Machine.MachineLineID == m_CurrentProduction.ProdMainMachineLine.MachineLineID).FirstOrDefault();
				if (MachineStandard != null)
				{
					if (MachineStandard.MinutesPerShift > 0)
					{
						m_MinPerHour = Convert.ToInt32(MachineStandard.MinutesPerShift / 7.16667);
					}
					else if (m_CurrentProduction.ProdMainItemID.MinutesPerShift > 0)
					{
						m_MinPerHour = Convert.ToInt32(m_CurrentProduction.ProdMainItemID.MinutesPerShift / 7.1667);
					}
					else
					{
						m_MinPerHour = MainXtraForm.MinPerHOur;
					}
				}
				else if (m_CurrentProduction.ProdMainItemID.MinutesPerShift > 0)
				{
					m_MinPerHour = Convert.ToInt32(m_CurrentProduction.ProdMainItemID.MinutesPerShift / 7.1667);
				}
				else
				{
					m_MinPerHour = MainXtraForm.MinPerHOur;
				}
			}
			else
			{
				m_MinPerHour = MainXtraForm.MinPerHOur;
			}

		}

	}
}